File: android/src/main/java/com/apparence/camerawesome/CameraPermissions.java
package com.apparence.camerawesome;

import android.app.Activity;
import android.content.pm.PackageManager;
import android.util.Log;

import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.StringJoiner;

import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.PluginRegistry;

import static android.Manifest.permission.CAMERA;
import static android.Manifest.permission.WRITE_EXTERNAL_STORAGE;

public class CameraPermissions implements EventChannel.StreamHandler, PluginRegistry.RequestPermissionsResultListener {

    private static final String TAG = CameraPermissions.class.getName();

    private static final List<String> permissions = new ArrayList<String>();

    private static final int PERMISSIONS_MULTIPLE_REQUEST = 5;

    private boolean permissionGranted = false;

    private EventChannel.EventSink events;



    public String[] checkPermissions(Activity activity, boolean ignoreExternalStorage) {
        if(activity == null) {
            throw new RuntimeException("NULL_ACTIVITY");
        }
        if (permissions.isEmpty()) {
            permissions.add(CAMERA);
            if (!ignoreExternalStorage) {
                permissions.add(WRITE_EXTERNAL_STORAGE);
            }
        }
        List<String> permissionsToAsk = new ArrayList<>();
        for(String permission : permissions) {
            if(ContextCompat.checkSelfPermission(activity, permission) != PackageManager.PERMISSION_GRANTED) {
                permissionsToAsk.add(permission);
            }
        }
        this.permissionGranted = permissionsToAsk.size() == 0;
        return permissionsToAsk.toArray(new String[0]);
    }

    public void checkAndRequestPermissions(Activity activity, boolean ignoreExternalStorage) {
        String[] permissionsToAsk = checkPermissions(activity, ignoreExternalStorage);
        if(permissionsToAsk.length > 0) {
            Log.d(TAG, "_checkAndRequestPermissions: " + String.join(",", permissionsToAsk));
            ActivityCompat.requestPermissions(
                    activity,
                    permissionsToAsk,
                    PERMISSIONS_MULTIPLE_REQUEST
            );
        }
    }

    public boolean hasPermissionGranted() {
        return permissionGranted;
    }

    // ---------------------------------------------
    // EventChannel.StreamHandler
    // ---------------------------------------------

    @Override
    public void onListen(Object arguments, EventChannel.EventSink events) {
        this.events = events;
    }

    @Override
    public void onCancel(Object arguments) {
        if(this.events != null) {
            this.events.endOfStream();
            this.events = null;
        }
    }

    // ---------------------------------------------
    // PluginRegistry.RequestPermissionsResultListener
    // ---------------------------------------------

    @Override
    public boolean onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        permissionGranted = true;
        for(int i=0; i < permissions.length; i++) {
            if(grantResults[i] != PackageManager.PERMISSION_GRANTED) {
                permissionGranted = false;
                break;
            }
        }
        if(this.events != null) {
            Log.d(TAG, "_onRequestPermissionsResult: granted " + String.join(", ", permissions));
            this.events.success(permissionGranted);
        } else {
            Log.d(TAG, "_onRequestPermissionsResult: received permissions but the EventSink is closed");
        }

        return permissionGranted;
    }
}

File: android/src/main/kotlin/com/apparence/camerawesome/CameraPermissions.kt
package com.apparence.camerawesome

import android.Manifest
import android.app.Activity
import android.content.pm.PackageManager
import android.util.Log
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.EventChannel.EventSink
import io.flutter.plugin.common.PluginRegistry.RequestPermissionsResultListener

class CameraPermissions : EventChannel.StreamHandler, RequestPermissionsResultListener {
    private var permissionGranted = false
    private var events: EventSink? = null
    fun checkPermissions(activity: Activity?): Array<String> {
        if (activity == null) {
            throw RuntimeException("NULL_ACTIVITY")
        }
        val permissionsToAsk: MutableList<String> = ArrayList()
        for (permission in permissions) {
            if (ContextCompat.checkSelfPermission(
                    activity,
                    permission
                ) != PackageManager.PERMISSION_GRANTED
            ) {
                permissionsToAsk.add(permission)
            }
        }
        permissionGranted = permissionsToAsk.size == 0
        return permissionsToAsk.toTypedArray()
    }

    fun checkAndRequestPermissions(activity: Activity?) {
        val permissionsToAsk = checkPermissions(activity)
        if (permissionsToAsk.size > 0) {
            Log.d(
                TAG,
                "_checkAndRequestPermissions: " + java.lang.String.join(",", *permissionsToAsk)
            )
            ActivityCompat.requestPermissions(
                activity!!,
                permissionsToAsk,
                PERMISSIONS_MULTIPLE_REQUEST
            )
        }
    }

    fun hasPermissionGranted(): Boolean {
        return permissionGranted
    }

    // ---------------------------------------------
    // EventChannel.StreamHandler
    // ---------------------------------------------
    override fun onListen(arguments: Any, events: EventSink) {
        this.events = events
    }

    override fun onCancel(arguments: Any) {
        if (events != null) {
            events!!.endOfStream()
            events = null
        }
    }

    // ---------------------------------------------
    // PluginRegistry.RequestPermissionsResultListener
    // ---------------------------------------------
    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ): Boolean {
        permissionGranted = true
        for (i in permissions.indices) {
            if (grantResults[i] != PackageManager.PERMISSION_GRANTED) {
                permissionGranted = false
                break
            }
        }
        if (events != null) {
            Log.d(
                TAG,
                "_onRequestPermissionsResult: granted " + java.lang.String.join(", ", *permissions)
            )
            events!!.success(permissionGranted)
        } else {
            Log.d(
                TAG,
                "_onRequestPermissionsResult: received permissions but the EventSink is closed"
            )
        }
        return permissionGranted
    }

    companion object {
        private val TAG = CameraPermissions::class.java.name
        private val permissions =
            arrayOf(
                Manifest.permission.CAMERA,
                Manifest.permission.WRITE_EXTERNAL_STORAGE,
                Manifest.permission.RECORD_AUDIO,
                Manifest.permission.ACCESS_FINE_LOCATION,
            )
        private const val PERMISSIONS_MULTIPLE_REQUEST = 5
    }
}